### PR TITLE
Improved PolyfaceData.ts

### PR DESCRIPTION
**Pull request by saeeedtorabi**
- Add missing checks to IndexedPolyface.terminateFacet
- Fixed few bugs
- Improved formatting and documentations of PolyfaceData.ts
- Added md doc for compress polyface data
- Added few tests

### CODE REVIEW

File: core/geometry/src/test/polyface/OffsetMeshContext.test.ts [hunk 1]
```diff
@@ -221,7 +221,7 @@ describe("OffsetMeshContext", () => {
     GeometryCoreTestIO.saveGeometry(allGeometry, "OffsetMeshContext", "OffsetDTM");
     expect(ck.getNumErrors()).equals(0);
   });
```
Comment (saeeedtorabi) [inline core/geometry/src/test/polyface/OffsetMeshContext.test.ts:223]: for got to remove ".only"
will remove them in the next commit  hopefully along with the fix for these 2 tests!

File: core/geometry/src/polyface/PolyfaceBuilder.ts [hunk 2]
```diff
@@ -112,8 +112,8 @@ class FacetSector {
    */
   public loadIndexedPointAndDerivativeCoordinatesFromPackedArrays(i: number, packedXYZ: GrowableXYZArray, packedDerivatives?: GrowableXYZArray, fractions?: GrowableFloat64Array, v?: number) {
     packedXYZ.getPoint3dAtCheckedPointIndex(i, this.xyz);
-    if (fractions && v !== undefined)
-      this.uv = Point2d.create(fractions.atUncheckedIndex(i), v);
+    if (fractions && v !== undefined) { }
+    // this.uv = Point2d.create(fractions.atUncheckedIndex(i), v);
```
Comment (saeeedtorabi) [inline core/geometry/src/polyface/PolyfaceBuilder.ts:116]: Dave please take a look at this.

Note that commenting out this line fixes the test. However, we still hit
```      
    if (messages.length > 0) {
        this.data.trimAllIndexArrays(lengthA);
        return messages;
      }
```
in `terminateFacet` with messages such as "Less than 3 indices in the last facet" and  "invalid point indices in the last facet".

File: core/geometry/src/polyface/PolyfaceData.ts [hunk 3]
```diff
@@ -451,27 +599,31 @@ export class PolyfaceData {
...
-  /** Reverse data in entire facet indexing arrays.
-   * * parameterized over type T so non-number data -- e.g. boolean visibility flags -- can be reversed.
+  /**
+   * Reverse data in an indexing array facet-by-facet for all of the facets specified by `facetStartIndex`.
+   * * Parameterized over type T so non-number data (e.g., boolean visibility flags) can be reversed.
+   * @param facetStartIndex start indices of all facets to be reversed.
+   * @param indices the indexing array, e.g., pointIndex, normalIndex, paramIndex, etc.
+   * @param preserveStart `true` to preserve the start of all facets (example: turn facet [1,2,3,4] to [1,4,3,2]);
+   * `false` to reverse all (example: turn facet [1,2,3,4] to [4,3,2,1]).
    */
```
Comment (saeeedtorabi) [inline core/geometry/src/polyface/PolyfaceData.ts:609]: reverseIndices and reverseIndicesSingleFacet functions call functions which are located at the end of this file so it's best to move this 2 functions to the end of this file for better code readability. However, I did not move them to make it easier to reveiw. I will move them in the next PR.

File: core/geometry/src/polyface/PolyfaceBuilder.ts (hunk 2 above)
Comment (dassaf4) [reply core/geometry/src/polyface/PolyfaceBuilder.ts]: Addressed!

**Review by dassaf4 — APPROVED** [review]
Made some comment suggestions, and swept in a bunch of polyface-related commits resulting from investigations inspired by this PR.

**Pull request merged**